src/libsyntax/attr.rs

```diff
@@ -312,6 +312,15 @@ pub fn find_export_name_attr(diag: &SpanHandler, attrs: &[Attribute]) -> Option<
     })
 }
 
+pub fn contains_extern_indicator(attrs: &[Attribute]) -> bool {
+    contains_name(attrs, "no_mangle") ||
+        contains_name(attrs, "export_name") ||
+        first_attr_value_str_by_name(attrs, "linkage").map(|v| match &*v {
```

Comment:
I think that we may want to tackle one issue at a time here rather than trying to dive into it all at once, let's separate out these linkage concerns for perhaps another PR?